### DISCUSSION TITLE
Add support for making a static website also on Windows?

### DISCUSSION THREAD

**Discussion by textplusdata** [datenstrom/community#150] [Ask a question] [answered]
Hi,

This follows on from other issues
- commandline build cannot write files? - Windows datenstrom/community#115
- Static pages not show 'error.html' when sth missing. datenstrom/community#116
- Better static build datenstrom/community#16
- Unusual pagination arguments datenstrom/community#65
- Enhance static website generator datenstrom/community#125
- Support for Windows Static Site build - anyone have a current workaround? datenstrom/community#149

I would like to see if I can try and code a workaround for windows users. However, my coding skills are pretty basic.

In the blog plugin the colon seems to be used in blog.php before the $key variable. e.g. for the published date or the tag summary

```
$output .= "<li><a href=\"".$blog->getLocation().$this->yellow->toolbox->normaliseArgs("published:$key")."\">"; 

$output .= "<li><a href=\"".$blog->getLocation().$this->yellow->toolbox->normaliseArgs("tag:$key")."\">";
```

Can anyone confirm if I'm right?

I'm not sure where the colon is implemented in the core.php file. Can anyone point me in the right direction for what to search for in the code?

Thanks!

**Comment by markseuffert** [comment]
Thanks for the suggestion, I think this feature request will impact many plugins.

A starting point for the [core](https://github.com/annaesvensson/yellow-core) is `class YellowToolbox`, search for `LocationArgs` and LOCATION_ARGS. For the other plugins I would search for occurrences of `normaliseArgs` and the `:` character.

Yellow is open source, help would be much appreciated. Shout my name if you need a code reviewer.  :smile:

**Comment by textplusdata** [comment]
Hi,
I'm a noob so this is too complicated for me. I can get the links to change in the html code from ":" to "--" for group separator for author/tags using the blog plugin but I can't find where the system is inserting the colon for the actual url.

I tried changing in core.php
   `$locationArgs .= "$key:$value";`
to
   `$locationArgs .= "$key"."--"."$value";`

But this seems to have no effect. I have tried replacing colons all over the place but I can't get anything to change.

I also tried 
`$locationArgs .= "$matches[1]"."--"."$matches[2]";`
and
`$locationArgs .= "$args[1]"."--"."$args[2]";`

under getlocationArgsNew 

If anyone can guess where I'm going wrong that would be very helpful

I should add that the broader plan is:
- add a boolean setting to config.ini called "WindowsMachine", default to 0
- add a variable to core called $groupSeparator
- default $groupSeparator to a colon. Change to "--" if WindowsMachine is true
-  use $groupSeparator in blog and other plugins for the separator e.g. author:name-of-author

**Comment by markseuffert** [comment, marked as answer]
You're right, this is not an easy feature at all. I tried a configuration setting at first, then I tried to detect Windows file system which resulted in less code. The important part is to use a separator that will not occur in regular locations, for example `:` and `=`. 

Please have a look at latest https://github.com/datenstrom/yellow. What do you think, would this work?

**Comment by wunderfeyd** [comment]
@textplusdata During the development I was able to test the new separator scheme on Windows by using Xampp. These tests included a static build with various standard plugins and the dynamic website generation. The blog is making strong usage of the url parameters and uses `.../author=test/` instead of `.../author:test` for example now (under Windows). All in all it was working fine, so we decided to tag a new version. Please download and test the latest Yellow. Your help is greatly appreciated.

**Comment by textplusdata** [comment]
Hi, 

This is great, thanks very much, and I'll take a look as soon as I can. 

One worry for me though is auto-detection for the separator. If you happened to move your development from a Windows machine to Linux/Mac, your links would change, which would obviously affect SEO/permalinks etc.

I think it would be more logical to set the desire for a windows-friendly separator in the config so that this could remain constant between development machines once it had been set. 

Error code for static site generation could tell Windows users to make sure they had config file set up correctly.

Will be back with more comments :-)

**Comment by BsNoSi** [comment]
Very nice. Thank you. This facilitates development/testing for me immensely because I work locally on Windows with XAMPP.

**Comment by markseuffert** [comment]
Windows build should be working, cross platform builds we have to tackle another time. Let us know if you run into any problems with the static site generator. Please open a new issue if there's more feedback.
